### PR TITLE
Make specache eagerly build - [MOD-4892]

### PR DESCRIPTION
**Pull request by GuyAv46**

### CODE REVIEW

File: src/spec.c [hunk 1]
```diff
@@ -1181,11 +1181,12 @@ static void IndexSpecCache_Free(IndexSpecCache *c) {
 // and at this point the refcount only gets decremented so there is no wory of some thread increasing the
 // refcount while we are freeing the cache.
 void IndexSpecCache_Decref(IndexSpecCache *c) {
-  if (!__atomic_sub_fetch(&c->refcount, 1, __ATOMIC_RELAXED)) {
+  if (c && !__atomic_sub_fetch(&c->refcount, 1, __ATOMIC_RELAXED)) {
```
Comment (meiravgri) [inline src/spec.c:1184]: in which case it becomes NULL?

File: src/spec.c [hunk 2]
```diff
@@ -1006,6 +1001,11 @@ static int IndexSpec_AddFieldsInternal(IndexSpec *sp, ArgsCursor *ac, QueryError
     }
     fs = NULL;
   }
+
+  // If we successfully modified the schema, we need to update the spec cache
+  IndexSpecCache_Decref(sp->spcache);
+  sp->spcache = IndexSpec_BuildSpecCache(sp);
+
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/spec.c:1008]: shouldn't we increase refcount of the new cache after building a new one?

File: src/spec.h [hunk 3]
```diff
@@ -356,6 +356,7 @@ IndexSpecCache *IndexSpec_GetSpecCache(const IndexSpec *spec);
 /**
  * Decrement the reference count of the spec cache. Should be matched
  * with a previous call of GetSpecCache()
+ * Can handle NULL
```
Comment (meiravgri) [inline src/spec.h:359]: why do we need to handle NULL?

**Review by meiravgri — COMMENTED** [review]
Good 🤩 
I think we need to increase the ref count (initialize to 1) upon building a new cache
Also, had a few comments regarding handling NULL (didn't see any scenario where it can happen)

File: src/rlookup.c [hunk 4]
```diff
@@ -39,7 +39,7 @@ static RLookupKey *createNewKey(RLookup *lookup, const char *name, size_t n, int
   return ret;
 }
 
-static RLookupKey *genKeyFromSpec(RLookup *lookup, const char *name, int flags) {
+static RLookupKey *genKeyFromSpec(RLookup *lookup, const char *name, size_t len, int flags) {
   const IndexSpecCache *cc = lookup->spcache;
   if (!cc) {
     return NULL;
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/rlookup.c:45]:
```suggestion
RS_LOG_ASSERT(spec->spcache, "Index spec cache is NULL");
```

[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/rlookup.c:45]: `Rlookup` doesn’t necessarily has a spec cache, it’s not an error we just skip this step

File: src/spec.c (hunk 1 above)
Comment (GuyAv46) [reply src/spec.c:1184]: All the places that can hold the cache today might not has one, so they all had a check if it’s not NULL before calling this. 
I removed the need to check if their possible cache is null or not

File: src/spec.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/spec.c:1008]: The cache is being built with refcount 1, which is the spec reference

File: src/spec.c [hunk 5]
```diff
@@ -2239,8 +2230,9 @@ int IndexSpec_CreateFromRdb(RedisModuleCtx *ctx, RedisModuleIO *rdb, int encver,
         sp->suffix = NewTrie(suffixTrie_freeCallback, Trie_Sort_Lex);
       }
     }
-
   }
+  // After loading all the fields, we can build the spec cache
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/spec.c:2234]: Yes

File: src/spec.h (hunk 3 above)
Comment (GuyAv46) [reply src/spec.h:359]: Instead of checking if it’s null before calling this

File: src/spec.c [hunk 6]
```diff
@@ -870,6 +870,8 @@ int IndexSpec_CreateTextId(const IndexSpec *sp) {
   return maxId + 1;
 }
 
+static IndexSpecCache *IndexSpec_BuildSpecCache(const IndexSpec *spec);
+
 /**
  * Add fields to an existing (or newly created) index. If the addition fails,
```
[the remark this answers is not in the document]
Comment (ephraimfeldblum) [reply src/spec.c:876]: We will never know... Only Mark knew and that was 5 years ago.

**Review by ephraimfeldblum — APPROVED** [review]
LGTM

**Pull request merged**